### PR TITLE
Events for exams IDA

### PR DESCRIPTION
**Pull request by ilee2u**
**Description:** We are developing a new backend for exams called edx-exams. This backend will be utilized by instructor dashboards to interact with student exam attempts, and will store data regarding those attempts.

We are currently working to implement the downstream effects triggered whenever an exam attempt is submitted or reviewed. For example, when an exam attempt is submitted, we will want to make sure edx-platform knows to mark the exam subsection.

We have added several event definitions based on different actions that can be performed on exam attempts:

- EXAM_ATTEMPT_SUBMITTED
- EXAM_ATTEMPT_REJECTED
- EXAM_ATTEMPT_VERIFIED
- EXAM_ATTEMPT_ERRORED
- EXAM_ATTEMPT_RESET

These events will be consumed by various services in edx-platform in order to perform tasks such as creating grade overrides, invalidating certificates, changing credit requirement statuses, and more.

**ISSUE:** [MST-2083](https://2u-internal.atlassian.net/browse/MST-2083) (Internal)

**Dependencies:** None.

**Merge deadline:** List merge deadline (if any)

**Installation instructions:** List any non-trivial installation
instructions.

**Testing instructions:**

1. Open page A
2. Do thing B
3. Expect C to happen
4. If D happened instead - check failed.

**Reviewers:**
- [ ] tag reviewer
- [ ] tag reviewer

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** List any concerns about this PR - inelegant
solutions, hacks, quick-and-dirty implementations, concerns about
migrations, etc.

### CODE REVIEW

File: openedx_events/learning/data.py
Comment (Zacharis278) [inline openedx_events/learning/data.py:314]: I noticed @MichaelRoytman's PR uses the UserData object here. Do you think we should use that?

File: openedx_events/learning/signals.py
Comment (Zacharis278) [inline openedx_events/learning/signals.py:214]: I'd leave off "by an instructor or staff" in case this no longer becomes true in the future. (AI review by an tool as an example)

File: openedx_events/learning/signals.py
Comment (Zacharis278) [inline openedx_events/learning/signals.py:225]: same as the above comment. Right now an LTI tool can automatically trigger this if there is no suspicious behavior.

**Review by Zacharis278 — COMMENTED** [review]
One comment on how we send user info. Otherwise this looks good to me

File: openedx_events/learning/data.py
Comment (ilee2u) [reply openedx_events/learning/data.py:314]: Looking at the parameters of the [downstream functions](https://2u-internal.atlassian.net/wiki/spaces/PT/pages/577699957/M6+Hackathon+Outline#(Old)-Parameters-of-downstream-functions-in-edx-platform%3A), it looks like most of them take user_id as a parameter anyways, so I think user_id is the way to go here.

Also I believe Michael's PR just has placeholder data for now.

Comment (MichaelRoytman) [reply openedx_events/learning/data.py:314]: I think it would make sense to use the `UserData`. It seems to be more common to use that than a `user_id`. But I do see uses of both, so there's precedence to use both.

I don't think we have to model our events off the way they're being consumed now. Ideally, they shouldn't be so context aware. I think, for me, it's less, "the Instructor Service currently only expects a `user_id` so that's what we should put in the event" and more "how can I best and most generically represent the concept of a user in my event"?

I guess the only consideration is maybe sending PII when we don't need to?

File: openedx_events/learning/signals.py
Comment (MichaelRoytman) [inline openedx_events/learning/signals.py:208]: Nit. I think `exam_attempt_data` could be a little verbose and `data` is a little redundant. What do you think about just `exam_attempt`? That matches the other events more closely.

Comment (ilee2u) [reply openedx_events/learning/signals.py:208]: Yeah that makes sense, changing now

File: openedx_events/learning/data.py
Comment (MichaelRoytman) [inline openedx_events/learning/data.py:318]: I thought we didn't need this anymore. Did something change?

Comment (ilee2u) [reply openedx_events/learning/data.py:318]: Nope just forgot to remove it

File: openedx_events/learning/data.py
Comment (Zacharis278) [reply openedx_events/learning/data.py]: I think there's a good chance future downstream logic would use the pii. This probably gets hooked up to name affirmation at some point.

**Review by Zacharis278 — APPROVED** [review]

**Review by MichaelRoytman — APPROVED** [review]
Nice job!

File: openedx_events/learning/data.py
Comment (robrap) [inline openedx_events/learning/data.py:311]: Docstring vs implementation for credit_requirement_status?

Comment (ilee2u) [reply openedx_events/learning/data.py:311]: Whoops! Fixing that now.

Comment (ilee2u) [reply openedx_events/learning/data.py:311]: For context, I removed that parameter and just forgot to remove the docstring.

File: openedx_events/learning/signals.py
Comment (robrap) [inline openedx_events/learning/signals.py:203]: Some high-level questions:
1. Does the platform have other types of "exams" that would not trigger these events? If so, how do we differentiate? Note that an event is defined by its event trigger + event data.
2. For the exams covered by these events, are there multiple exam types? Would it be useful to know that as part of the event data?
3. If there are different exam types, do some of these events only apply to certain types of exam types? If so, would that be useful to note?

Comment (Zacharis278) [reply openedx_events/learning/signals.py:203]: 1. kind of. The existing exam system (edx-proctoring) does not trigger these events, the service that emits these is intended to replacement. There are no other types of exams outside that.
2. Good point. There are multiple exam types and edx-platform is _mostly_ unaware of this since all the state/ui flows are manage by the exams/proctoring service. That said, I think there is value including this. Consumers may likely care if an exam is proctored or not. We might have actually missed a need for this. I think Credits depend on if an exam is proctored.
3. Yup some of these events are impossible depending on the exam type. A timed exam cannot be rejected since there is no video review to reject for example. We should add those details to the event descriptions

Comment (robrap) [reply openedx_events/learning/signals.py:203]: Are there non-timed exams, and are those included in this as well? Or are these just timed-exams, proctored and not-proctored? Just curious.

Comment (Zacharis278) [reply openedx_events/learning/signals.py:203]: 🤔  I guess this doesn't include 'all' exams just 'special exams' which have a different UX and workflow. Right now, all special exam types are a subset of timed. This wouldn't include normal sections instructors are using as or are graded as exams however. We could possibly reintroduce the term 'special exam' for these events which could clear this up.

Comment (robrap) [reply openedx_events/learning/signals.py:203]: If you think of this more as an API discussion around exams and special-exams, who do you think should be a part of this discussion? Maybe https://openedx.atlassian.net/wiki/spaces/COMM/pages/3554214289/Arch+Coordination+-+How+to+announce+and+tune-in will help you get the right discussion going?

Comment (robrap) [reply openedx_events/learning/signals.py:203]: Related: There are suggestions https://docs.openedx.org/projects/openedx-events/en/latest/how-tos/creating-new-events.html#propose-the-new-event-to-the-community for how to propose these new events.

Comment (ilee2u) [reply openedx_events/learning/signals.py:203]: Hi Robert!

To answer your question, you could technically title a course subsection as an "Exam" and grade it as an exam, however our exams backend (i.e. the exams IDA we're developing, edx-exams) would never know about it, as the exam backend is only used/called when a subsection is marked as a "Special Exam", i.e. a timed or proctored exam.

As such, the events that we are defining would only ever be produced in the context of a timed or proctored exam. The logic that we're implementing to produce these events mirrors the logic that edx-proctoring uses to call functions in various edx-platform services (e.g. calling the credit service to set a leaner's credit requirement status).

The bulk of the logic that we're porting from edx-proctoring can be seen here: https://github.com/openedx/edx-proctoring/blob/8091977772ec8967c79e854de7e0c813ee7c3d7e/edx_proctoring/api.py#L1479 (I recommend viewing this in an IDE since it can go a little deep).

Since certain exam attempt statuses (i.e. verified, rejected, and errored) are inherently linked to proctored exams, this means that edx-proctoring can distinguish if an exam is proctored or not based on the incoming status to update the attempt to (Side note that we never have a case where we need to know if an exam is specifically timed and not proctored).

Since we are simply mirroring this behavior in the exams IDA, this means that events produced in the exams IDA would be consumed by the intended services naturally, without us needing to gate anything behind a if exam_type=='proctored' kind of statement.

The pseudocode of how we'll mimic edx-proctoring's logic in edx-exams can be seen in an internal document that I can DM to you.

With all this being said, I agree that there could be value in attaching the exam type to the event down the line.

Comment (robrap) [reply openedx_events/learning/signals.py:203]: > We could possibly reintroduce the term 'special exam' for these events which could clear this up.

Something like this sounds reasonable, but the point of my last two messages is that it makes sense to bring in some other domain experts to help decide this type of question.

Comment (robrap) [reply openedx_events/learning/signals.py:203]: @ilee2u: I posted my last comment at the same time you were posting yours, but it applies equally to you. This all sounds reasonable, but it would be great to get others involved who work in this area to help with names and details to make the events more clear and forward thinking.

Comment (Zacharis278) [reply openedx_events/learning/signals.py:203]: @ilee2u I think a discuss post referencing this PR in combination with the [original ADR](https://github.com/openedx/edx-proctoring/pull/1049) for the project would be a good start to get input

Comment (ilee2u) [reply openedx_events/learning/signals.py:203]: A Discuss post for this topic has been created here: https://discuss.openedx.org/t/how-to-implement-events-for-the-exams-ida/11276

Comment (robrap) [reply openedx_events/learning/signals.py:203]: Can you confirm that you agreed/decided that no name or doc changes were needed to clarify what is meant by "exam"? It feels like at least a doc update might be warranted, unless you think I'm the only person that would ever ask about this.

Comment (mariajgrimaldi) [reply openedx_events/learning/signals.py:203]: Hello folks. I do believe we should use a different name for the event. `exam` might mean something in `edx-exams`, but it could mean other (similar) things in `edx-platform` that need another type of data for the event. What do you think?

Comment (Zacharis278) [reply openedx_events/learning/signals.py:203]: ^ I think this is a good idea to have before we merge. I could see people getting confused on two points

1. these events only pertain to 'special' exams
2. these events are not published by the existing exams system

File: openedx_events/learning/data.py
Comment (mariajgrimaldi) [inline openedx_events/learning/data.py:311]: Nit:
```suggestion
        student_user (UserData): user object for the student to whom the exam attempt belongs
        course_key (CourseKey): identifier of the course to which the exam attempt belongs
        usage_key (UsageKey): identifier of the content that will get a exam attempt
        exam_type (str): type of exam that was taken (e.g. timed, proctored, etc.)
        requesting_user (UserData): user triggering the event (sometimes a non-learner, e.g. an instructor)
```

**Review by mariajgrimaldi — COMMENTED** [review]
Hi there folks. I left a few minor comments. Thanks!

File: openedx_events/learning/signals.py
Comment (ilee2u) [reply openedx_events/learning/signals.py]: I've made the following changes to address this:

- Changed the data name to "SpecialExamAttemptData"
- Added a comment about how events using this data are only emitted for special exams from the newer exams backend, edx-exams

Comment (mariajgrimaldi) [reply openedx_events/learning/signals.py]: That's better! Thank you folks.

File: CHANGELOG.rst
Comment (mariajgrimaldi) [inline CHANGELOG.rst:20]:
```suggestion
[8.7.0] - 2023-09-29
```

Comment (ilee2u) [reply CHANGELOG.rst]: Done 👍

**Comment by ilee2u** [comment]
Hi, we've been informed that v8.6.0, which this PR is based off, is having issues, so please do not merge this until we find a solution/workaround for those issues.

File: openedx_events/learning/signals.py
Comment (robrap) [reply openedx_events/learning/signals.py]: I don't want to push hard on this, because it's not my domain of expertise, but the naming difference between EXAM_ATTEMPT_SUBMITTED and SpecialExamAttemptData just catches my attention. The obvious other option is to rename to SPECIAL_EXAM_ATTEMPT_SUBMITTED , and I guess the question I have is around the pros/cons of going with the more specific name.

File: openedx_events/learning/data.py
Comment (robrap) [inline openedx_events/learning/data.py:306]: Do we want another point, like?:
```
* Use the prefix `EXAM_`, which is equivalent to "special exam", but the shorter name is to hopefully avoid confusion because these are the events and exams that would likely be of concern.
```
Feel free to reword.

File: openedx_events/learning/signals.py
Comment (Zacharis278) [reply openedx_events/learning/signals.py]: 🤔 🤔 It's a good question, I'd almost lean towards not using 'special' in the name at all and just commenting to the meaning? Otherwise, we need to consistently use the term across Open edX. I don't think 'special exam' has any additional meaning to anyone working outside this domain unless it's described. There was a actually discussion a while back that we stop using this term because it's just confusing, trying to dig that up.

I'm not sure how to best describe the fact that this intends to include all content that is treated with an exam user experience, but not content that might be titled (on the section or grading policy) by course teams as 'exam'.  The latter is ultimately no different from other content, the platform doesn't do anything functionally different. (hence not special)

File: docs/decisions/0013-special-exam-submission-and-review-events.rst
Comment (Zacharis278) [inline docs/decisions/0013-special-exam-submission-and-review-events.rst:19]: I think this isn't all that important from the perspective of someone working with events.

File: docs/decisions/0013-special-exam-submission-and-review-events.rst
Comment (Zacharis278) [inline docs/decisions/0013-special-exam-submission-and-review-events.rst:9]: I think the context section might be a good place to explain the meaning of exam / special exam.

Rough points to hit
- course sections with an `exam_type` have additional logic that governs completion, grading, credit requirements, and more based on the type of exam.
- these sections are sometimes referred to as special exams
- course sections without an exam type configured may still be graded as an exam. This type of content does not have the exam user experience and is not governed by any exam specific logic.

File: docs/decisions/0013-special-exam-submission-and-review-events.rst
Comment (Zacharis278) [inline docs/decisions/0013-special-exam-submission-and-review-events.rst:34]: nit: 
```suggestion
* `edx-platform` will consume these events in order to handle all behavior as it pertains to the state of an exam subsection.
```

**Comment by mariajgrimaldi** [comment]
Hi @ilee2u! This PR introduced an ADR with the status draft. Do you think we can change the ADR's status to accepted by now?

**Comment by ilee2u** [comment]
> > Hi @ilee2u! This PR introduced an ADR with the status draft. Do you think we can change the ADR's status to accepted by now?
> 
> Hi again, yes we should be able to! Thanks :)

Just made a PR: https://github.com/openedx/openedx-events/pull/410

**Pull request merged**
